cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -277,21 +277,32 @@ func getSchedulerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
 }
 
 // getProxyEnvVars builds a list of environment variables to use in the control plane containers in order to use the right proxy
-func getProxyEnvVars() []v1.EnvVar {
+func getProxyEnvVars(cfg *kubeadmapi.MasterConfiguration) []v1.EnvVar {
+	noProxyIPs := getNoProxyIPs(cfg)
 	envs := []v1.EnvVar{}
 	for _, env := range os.Environ() {
 		pos := strings.Index(env, "=")
 		if pos == -1 {
 			// malformed environment variable, skip it.
 			continue
 		}
-		name := env[:pos]
+		name := strings.ToLower(env[:pos])
 		value := env[pos+1:]
-		if strings.HasSuffix(strings.ToLower(name), "_proxy") && value != "" {
+
+		if strings.HasSuffix(name, "_proxy") && value != "" {
+			if name == "no_proxy" {
+				noProxyIPs = append(noProxyIPs, value)
+				continue
+			}
 			envVar := v1.EnvVar{Name: name, Value: value}
 			envs = append(envs, envVar)
 		}
 	}
+
+	// set no_proxy
+	envVar := v1.EnvVar{Name: "no_proxy", Value: strings.Join(noProxyIPs, ",")}
```

Comment:
This could be a really long list if you have a large CIDR block. Isn't there a limit on how long an environment variable can be (less than 32k)?

Reply:
There is limit. And transport connections supports CIDR notation of subnets in no_proxy.
There is no need to expand them at all.

Reply:
@kad AFAIK, golang does not support no_proxy with CIDR notation of subnet. 
Here is an on-going issue golang/go#16704 on this.

> Isn't there a limit on how long an environment variable can be (less than 32k)
> There is limit.

Yeah. https://www.in-ulm.de/~mascheck/various/argmax/ lists Actual values for `ARG_MAX`. This is a risk when setting environments.

Reply:
That is exactly the reason why I consider that patch is incorrect:
1. kubeadm and other kubernetes components are utilizing modified http transport which supports CIDR notations.
2. kubeadm gives warnings during init stage if access to any IP address in pod or service CIDRs will be going via proxies, according to values of *_proxy variables.
3. If system admin require traffic to be going via proxies, he is able to ignore warning.
4. if system admin wants to have traffic going directly = values of environment variables before running kubeadm should be including pod and service cidrs.
5. with current idea of the patch behaviour becomes inconsistent: kubeadm will produce warning, but anyway will alter variables put into manifests. That is less predictable for sysadmins and will cause more issues.
6. trying to expand IPv6 pod/service cidrs will lead to OOM-kill of kubeadm.

So, to summarise: if in issue  kubernetes/kubeadm#666 warning was correctly shown to user of kubeadm but was ignored by user, we don't need this PR at all. If for some reason warning was not shown, we need to investigate more.

Reply:
@kad Thanks for the explanation.

> And transport connections supports CIDR notation of subnets in no_proxy.
There is no need to expand them at all.

I did found the link https://github.com/kubernetes/kubernetes/blob/f31ac9e0b6e70516648970699fe04f929a7a320e/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L86-L88 .

Will update it.